twsf/mastodon/toots.js

```diff
@@ -0,0 +1,176 @@
+require('dotenv').config();
+
+const fetch = require('node-fetch');
+const logger = require('../../logger');
+const {addNewGuess, guessTypes} = require('../../database/twsf');
+
+// const headers = {
+//     Authorization: `Bearer ${process.env.MASTODON_TOKEN}`,
+// };
+
+/*
+const fetchTwsfTootHistory = async (instance) => {
+    const path = '/api/v2/search';
+    const url = new URL(instance + path);
+
+    url.searchParams.append('q', 'thisweeksf');
+    url.searchParams.append('type', 'hashtag');
+
+    const resp = await fetch(url);
+    if (!resp.ok) {
+        console.error(
+            `Encountered ${resp.statusText} when fetching peers from ${instance}`,
+        );
+        return [];
+    }
+
+    const jsn = await resp.json();
+    return jsn.hashtags.find((ht) => ht.name === 'thisweeksf').history;
+};
+*/
+
+const tootIsRemote = (status) => {
+    status.account.acct.includes('@');
+};
+const getTootInstanceName = (status) => {
+    return status.url.match(/https?:\/\/(.*?)\//)[1];
+};
+const normalizeAcct = (status) => {
+    if (tootIsRemote(status)) return status.account.acct;
+    return status.account.username + '@' + getTootInstanceName(status);
+};
+
+const fetchSearchPage = async (url) => {
+    const resp = await fetch(url);
+    if (!resp.ok) {
+        console.error(`Encountered ${resp.statusText} when fetching ${url}`);
+        return {statuses: []};
+    }
+
+    const statuses = await resp.json();
+
+    const links = resp.headers.get('Link');
+    const next = links?.match(/rel="next", <(.*?)>/)?.[1];
+
+    return {statuses, next};
+};
+
+const fetchTwsfToots = async (instance, sinceId) => {
+    const path = '/api/v1/timelines/tag/thisweeksf';
+    const url = new URL(instance + path);
+    if (sinceId) url.searchParams.append('since_id', sinceId);
+
+    const fetchedToots = [];
+    let statuses;
+    let next = url;
+
+    while (next) {
+        ({statuses, next} = await fetchSearchPage(next));
+        fetchedToots.push(...statuses);
+    }
+
+    return fetchedToots;
+};
+
+const getTootReplies = async (status) => {
+    const instance = 'http://' + getTootInstanceName(status);
+    const path = `/api/v1/statuses/${status.id}/context`;
+    const url = instance + path;
+
+    const resp = await fetch(url);
+    if (!resp.ok) {
+        console.error(
+            `Encountered ${resp.statusText} when fetching replies to ${status.id}`,
+        );
+        return [];
```

Comment:
as this is not a Promise, you may get an error in the caller if it ever happens
(I think async / await may be really nice to manage these kind of things so I may be wrong)